### PR TITLE
Add missing datatype functions to new API

### PR DESCRIPTION
**Pull request by ajreynol**
This is in preparation for migrating the parser to use the Term-level API for datatypes.

Notably, this adds the function `mkDatatypeSorts` for making mutually recursive datatypes.  I've added a unit test that demonstrates this method (which mirrors the Expr-level datatype API).

### CODE REVIEW

File: src/api/cvc4cpp.cpp [hunk 1]
```diff
@@ -2458,6 +2469,33 @@ Term Solver::mkTermInternal(Kind kind, const std::vector<Term>& children) const
...
 }
 
+std::vector<Sort> Solver::mkDatatypeSortsInternal(
+    std::vector<DatatypeDecl>& dtypedecls,
+    std::set<Sort>& unresolvedSorts) const
+{
+  CVC4_API_SOLVER_TRY_CATCH_BEGIN;
+
+  std::vector<CVC4::Datatype> datatypes;
+  for (unsigned i = 0, ndts = dtypedecls.size(); i < ndts; i++)
```
Comment (aniemetz) [inline src/api/cvc4cpp.cpp]: Use `size_t` for `i`, `ndts`.

File: test/unit/api/datatype_api_black.h [hunk 2]
```diff
@@ -57,6 +58,77 @@ void DatatypeBlack::testMkDatatypeSort()
...
+    TS_ASSERT(dtsorts[i].isDatatype());
+    TS_ASSERT(!dtsorts[i].getDatatype().isFinite());
+    TS_ASSERT(dtsorts[i].getDatatype().getName() == dtdecls[i].getName());
+  }
+  // verify the resolution was correct
+  Datatype dtTree = dtsorts[0].getDatatype();
+  DatatypeConstructor dtcTreeNode = dtTree[0];
+  TS_ASSERT(dtcTreeNode.getName() == "node");
+  DatatypeSelector dtsTreeNodeLeft = dtcTreeNode[0];
+  TS_ASSERT(dtsTreeNodeLeft.getName() == "left");
```
Comment (aniemetz) [inline test/unit/api/datatype_api_black.h]: Add `TS_ASSERT_THROWS` check for `getName()`.

Comment (ajreynol) [reply test/unit/api/datatype_api_black.h]: I can add a TS_ASSERT_THROWS_NOTHING (DatatypeDecl::getName can never throw anything, it should always have a non-null datatype by construction).

Comment (aniemetz) [reply test/unit/api/datatype_api_black.h]: Ah sorry, that was the wrong line. I meant to write this for this:https://github.com/CVC4/CVC4/pull/3930/files#diff-d4eff0da8e72b4f2d088f43ce4ecf899R1883

File: test/unit/api/datatype_api_black.h [hunk 3]
```diff
@@ -238,4 +238,7 @@ void DatatypeBlack::testDatatypeNames()
   DatatypeSelector dselTail = dcons[1];
   TS_ASSERT(dselTail.getName() == std::string("tail"));
   TS_ASSERT(dselTail.getRangeSort() == dtypeSort);
+
+  // possible to construct null datatype declarations if not using solver
+  TS_ASSERT_THROWS_NOTHING(DatatypeDecl().getName());
```
Comment (aniemetz) [inline test/unit/api/datatype_api_black.h]: This should actually throw.

**Pull request merged**